numpy/ctypeslib.py

```diff
@@ -287,10 +290,23 @@ def ndpointer(dtype=None, ndim=None, shape=None, flags=None):
             except Exception:
                 raise TypeError("invalid flags specification")
             num = _num_fromflags(flags)
+
+    # normalize shape to an Optional[tuple]
+    if shape is not None:
+        try:
+            shape = tuple(shape)
+        except TypeError:
+            # single integer -> 1-tuple
+            shape = (shape,)
+
+    cache_key = (dtype, ndim, shape, num)
```

Comment:
I wonder if #1725 might be a problem here, AFAIK, dtypes are not immutable, although the topic has come up many times and I don't recall where it has ended up.

Reply:
Yep, it will be. I'm of the opinion that dtypes need to become immutable